### PR TITLE
WINDUPRULE-126 new Resteasy rules for EAP7 migration

### PR DESCRIPTION
**Pull request by mareknovotny**
https://issues.jboss.org/browse/WINDUPRULE-126

### CODE REVIEW

**Comment by mareknovotny** [comment]
There are still missing migration path  for some deprecated classes/interfaces, but the raw rule structures are there. I will enhance hint messages as a next step.

**Comment by Ladicek** [comment]
Sorry for not getting to this yet, I had too much other work. I hope to get to it next week.

File: rules-reviewed/eap7/eap6/resteasy.windup.xml [hunk 1]
```diff
@@ -0,0 +1,826 @@
...
+                </hint>
+            </perform>
+        </rule>
+        <rule id="resteasy-eap6-000007">
+            <when>
+                <javaclass references="org.jboss.resteasy.client.ClientRequestFactory" />
+            </when>
+            <perform>
+                <hint title="Deprecated ClientRequestFactory class" effort="1" severity="mandatory">
+                    <message>Replace the `org.jboss.resteasy.client.ClientRequestFactory` usage with `org.jboss.resteasy.client.jaxrs.org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder` class.</message>
```
Comment (Ladicek) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: `org.jboss.resteasy.client.jaxrs.org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder` is definitely not right :-)

**Comment by Ladicek** [comment]
Correct spelling is `RESTEasy`. The rules seem to use `RESTeasy`, which is wrong (though upstream doesn't seem to be very consistent either :-) ).

**Comment by Ladicek** [comment]
Rules `resteasy-eap6-000008` and `resteasy-eap6-000101` catch the same class. Better merge them into one (`resteasy-eap6-000008` seems to have better description, `resteasy-eap6-000101` has more links).

File: rules-reviewed/eap7/eap6/resteasy.windup.xml [hunk 2]
```diff
@@ -0,0 +1,826 @@
...
+                <hint title="Deprecated ClientExecutionInterceptor interface" effort="1" severity="mandatory">
+                    <message>Use ### instead of `org.jboss.resteasy.spi.interception.ClientExecutionInterceptor`.
+                    The Resteasy interceptor facility introduced in release 2.x is replaced by the JAX-RS 2.0 compliant interceptor facility in release 3.0.x.</message>
+                    <link title="RESTeasy SPI Application Changes"
+                        href="https://access.qa.redhat.com/documentation/en/red-hat-jboss-enterprise-application-platform/version-7.0/migration-guide/#migrate_resteasy_deprecated_classes" />
+                    <tag>resteasy</tag>
+                </hint>
+            </perform>
+        </rule>
+        <rule id="resteasy-eap6-000024">
```
Comment (Ladicek) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: Duplicates `resteasy-eap6-000015` and uses wrong FQN in description. Should IMHO be deleted.

File: rules-reviewed/eap7/eap6/resteasy.windup.xml [hunk 3]
```diff
@@ -0,0 +1,826 @@
...
+                    <tag>resteasy</tag>
+                </hint>
+            </perform>
+        </rule>
+        <rule id="resteasy-eap6-000030">
+            <when>
+                <javaclass references="org.jboss.resteasy.spi.interception.PreProcessInterceptor" />
+            </when>
+            <perform>
+                <hint title="Deprecated MessageBodyWriterInterceptor interface" effort="1" severity="mandatory">
```
Comment (Ladicek) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: Wrong `<hint title="...">` and `<message>` -- this rule is about `PreProcessInterceptor`

File: rules-reviewed/eap7/eap6/resteasy.windup.xml [hunk 4]
```diff
@@ -0,0 +1,826 @@
...
+                <hint title="RESTeasy SPI Application Change for StringConverter" effort="1" severity="mandatory">
+                    <message>Replace `org.jboss.resteasy.spi.StringConverter` by `javax.ws.rs.ext.ParamConverterProvider`.</message>
+                    <link title="RESTeasy SPI Application Changes"
+                        href="https://access.redhat.com/documentation/en/red-hat-jboss-enterprise-application-platform/7.0.beta/migration-guide/chapter-4-application-changes" />
+                    <link title="JAX-RS API" href="https://jax-rs-spec.java.net/nonav/2.0-rev-a/apidocs/index.html?javax/ws/rs/ext/ParamConverterProvider.html" />
+                    <tag>resteasy</tag>
+                </hint>
+            </perform>
+        </rule>
+        <rule id="resteasy-eap6-000010">
```
Comment (Ladicek) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: Similar rule should exist for `org.jboss.resteasy.spi.Registry`. They are mentioned together in the Migration Guide.

**Comment by Ladicek** [comment]
I deleted my previous comments about duplicate rules because I found that there's more. Here's a list:

- `resteasy-eap6-000008` and `resteasy-eap6-000101`
- `resteasy-eap6-000011` and `resteasy-eap6-000028`
- `resteasy-eap6-000012` and `resteasy-eap6-000026`
- `resteasy-eap6-000013` and `resteasy-eap6-000027`
- `resteasy-eap6-000015` and `resteasy-eap6-000024`
- `resteasy-eap6-000018` and `resteasy-eap6-000031`

**Comment by Ladicek** [comment]
In addition to a missing rule for `org.jboss.resteasy.spi.Registry`, which I already mentioned in a comment above, there are rules missing for `org.jboss.resteasy.spi.ResteasyProviderFactory` and `org.jboss.resteasy.annotations.cache.ServerCached`. Both of them are documented in https://access.qa.redhat.com/documentation/en/red-hat-jboss-enterprise-application-platform/version-7.0/migration-guide/#migrate_removed_or_protected_resteasy_classes

File: rules-reviewed/eap7/eap6/resteasy.windup.xml [hunk 5]
```diff
@@ -0,0 +1,826 @@
...
+                    <link title="RESTeasy SPI Application Changes"
+                        href="https://access.qa.redhat.com/documentation/en/red-hat-jboss-enterprise-application-platform/version-7.0/migration-guide/#migrate_resteasy_deprecated_classes" />
+                    <link title="Resteasy MultipartFormDataInput#getFormData API" href="https://docs.jboss.org/resteasy/docs/3.0.13.Final/javadocs/org/jboss/resteasy/plugins/providers/multipart/MultipartFormDataInput.html#getFormData()" />
+                    <tag>resteasy</tag>
+                </hint>
+            </perform>
+        </rule>
+        <rule id="resteasy-eap6-000124">
+            <when>
+                <javaclass references="Acme.Serve.Serve.AcmeSession.getSessionContext{*}">
```
Comment (Ladicek) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: I don't know what this is but I'm 99.999 % sure that this has never been shipped nor supported as part of EAP. (To confirm, I unpacked all RESTEasy JARs from EAP 6.4.0 and indeed, these classes are nowhere to be found.)

File: rules-reviewed/eap7/eap6/resteasy.windup.xml [hunk 6]
```diff
@@ -0,0 +1,826 @@
...
+                    <link title="RESTeasy SPI Application Changes"
+                        href="https://access.qa.redhat.com/documentation/en/red-hat-jboss-enterprise-application-platform/version-7.0/migration-guide/#migrate_resteasy_deprecated_classes" />
+                    <link title="JAX-RS API" href="https://docs.oracle.com/javaee/7/api/index.html?javax/ws/rs/ext/ReaderInterceptor.html" />
+                    <tag>resteasy</tag>
+                </hint>
+            </perform>
+        </rule>
+        <rule id="resteasy-eap6-000017">
+            <when>
+                <javaclass references="org.jboss.resteasy.client.core.executors.ApacheHttpClientExecutor" />
```
Comment (Ladicek) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: For some reason, this class has never been shipped with EAP 6, even though it exists in upstream repositories of the corresponding versions. Oh well :-(

**Comment by Ladicek** [comment]
Also, the hints seem to be quite inconsistent. For example, there are 2 rules for deprecated methods on the `ServerCookie` class:

- `resteasy-eap6-000118`: Deprecated ServerCookie#checkName method
- `resteasy-eap6-000127`: ServerCookie#maybeQuote method deprecation

But this happens in more rules. Unifying a bit would be great IMHO.

**Comment by Ladicek** [comment]
OK, so I developed tests for everything described in the Migration Guide except the SPI exceptions. That seems to be cca 1/2 to 2/3 of the rules that are in this PR. I will continue next week, but all the remaining rules (besides the SPI exceptions) look more like corner cases or classes that might have never been shipped with EAP.

File: rules-reviewed/eap7/eap6/resteasy.windup.xml (hunk 1 above)
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: good catch :)

File: rules-reviewed/eap7/eap6/resteasy.windup.xml (hunk 4 above)
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: I missed that before in the doc, I added *resteasy-eap6-000032*

File: rules-reviewed/eap7/eap6/resteasy.windup.xml (hunk 5 above)
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: you are right, I just took everything what is deprecated from javadoc documentation, so there could be such cases.

**Comment by mareknovotny** [comment]
I updated everything except 

> But this happens in more rules. Unifying a bit would be great IMHO.

There are few unclear things to me which i sent a request email to resteasy team to answer.

**Comment by Ladicek** [comment]
When asking for "unification", I meant the wording. Using the example again:

- `resteasy-eap6-000118`: Deprecated ServerCookie#checkName method
- `resteasy-eap6-000127`: ServerCookie#maybeQuote method deprecation

There's really no reason why these two should use different wording. But I agree that this is minor.

**Comment by mareknovotny** [comment]
ok, i got these two and I "unified" them ;), See in code https://github.com/windup/windup-rulesets/pull/128/commits/69bd9589887304f29ed19f0d27216878ac58c6a4#diff-3a443b762e1e02a9bd95cd1e7356192cR341

**Comment by Ladicek** [comment]
Nitpicking: the RESTEasy project name spelling is still wrong on some places. It's `RESTEasy`, not `RESTeasy` nor `Resteasy`.

**Comment by Ladicek** [comment]
A majority of rules related to interceptors use a hint of the form "Deprecated XXX interface". See e.g. `resteasy-eap6-000011`, `resteasy-eap6-000014`, `resteasy-eap6-000015`, `resteasy-eap6-000019` etc.

There are two exceptions: rules `resteasy-eap6-000012` and `resteasy-eap6-000013` instead use "RESTeasy SPI Application Changes for XXX". I believe that rules `resteasy-eap6-000012` and `resteasy-eap6-000013` should be changed to the common form of "Deprecated XXX interface". (Or the other way around, but this seems more natural to me.)

**Comment by mareknovotny** [comment]
ok, i finished it and while waiting on RESTEasy ;) team to confirm my questions from my email, all others seem done.

File: rules-reviewed/eap7/eap6/resteasy.windup.xml [hunk 7]
```diff
@@ -0,0 +1,744 @@
...
+                    <link title="RESTEasy SPI Registry Changes"
+                        href="https://docs.jboss.org/resteasy/docs/3.0.13.Final/javadocs/index.html?org/jboss/resteasy/spi/Registry.html" />
+                    <tag>resteasy</tag>
+                </hint>
+            </perform>
+        </rule>        
+        <!-- Exceptions -->
+        <rule id="resteasy-eap6-000101">
+            <when>
+                <javaclass references="org.jboss.resteasy.spi.ForbiddenException" />
```
Comment (Ladicek) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: This seems to only exist in RESTEasy 3. But it's true that it's deprecated, so I think we can keep the rule.

**Comment by Ladicek** [comment]
The rules for the JAXB provider plugin (`resteasy-eap6-000119`, `resteasy-eap6-000126`, `resteasy-eap6-000128` and `resteasy-eap6-000130`) report these warnings:

```
Deprecated method JAXBContextWrapper#createValidator
Deprecated SecureUnmarshaller#isValidating
SecureUnmarshaller#setAdapter method deprecation
SecureUnmarshaller#setValidating method deprecation
```

Again, the wording should be unified. Also, the first warning is "mandatory", the remaining 3 are "optional". Not sure why.

**Comment by Ladicek** [comment]
OK, so I finished writing tests for all the rules. Here's a list of classes that have never been shipped with EAP 6 (checked with EAP 6.0.0 and 6.4.0) and the list of corresponding rules -- I didn't write tests for those, but I suppose they can stay:

- `org.jboss.resteasy.client.core.executors.ApacheHttpClientExecutor`: `resteasy-eap6-000017`
- `org.jboss.resteasy.plugins.cache.server.SimpleServerCache`: `resteasy-eap6-000018` 
- `org.jboss.resteasy.core.ResteasyHttpServletResponseWrapper`: `resteasy-eap6-000120`, `resteasy-eap6-000121`, `resteasy-eap6-000125`, `resteasy-eap6-000129`
- `org.jboss.resteasy.auth.oauth.OAuthValidator`: `resteasy-eap6-000131`

**Comment by mareknovotny** [comment]
i hope i unified all rule titles. Check it out.

File: rules-reviewed/eap7/eap6/resteasy.windup.xml [hunk 8]
```diff
@@ -0,0 +1,743 @@
...
+                    <tag>jax-rs</tag>
+                </hint>
+            </perform>
+        </rule>
+        <rule id="resteasy-eap6-000107">
+            <when>
+                <javaclass references="org.jboss.resteasy.spi.UnsupportedMediaTypeException" />
+            </when>
+            <perform>
+                <hint title="Deprecated class UnsupportedMediaTypeException in RESTEasy 3" effort="1" severity="optional">
```
Comment (Ladicek) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: I guess you want to say "Deprecated exception UnsupportedMediaTypeException in RESTEasy 3", similarly to other deprecated exceptions.

File: rules-reviewed/eap7/eap6/resteasy.windup.xml [hunk 9]
```diff
@@ -0,0 +1,743 @@
...
+                    <tag>resteasy</tag>
+                </hint>
+            </perform>
+        </rule>
+        <rule id="resteasy-eap6-000013">
+            <when>
+                <javaclass references="org.jboss.resteasy.spi.interception.MessageBodyWriterContext" />
+            </when>
+            <perform>
+                <hint title="Deprecated interface MessageBodyWriterContext" effort="1" severity="optional">
```
Comment (Ladicek) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: I guess you might want to add "in RESTEasy 3" similarly to other rules.

File: rules-reviewed/eap7/eap6/resteasy.windup.xml [hunk 10]
```diff
@@ -0,0 +1,743 @@
...
+            </perform>
+        </rule>
+        <rule id="resteasy-eap6-000119">
+            <when>
+                <javaclass references="org.jboss.resteasy.plugins.providers.jaxb.JAXBContextWrapper.createValidator{*}">
+                    <location>METHOD_CALL</location>
+                </javaclass>
+            </when>
+            <perform>
+                <hint title="Deprecated method JAXBContextWrapper#createValidator" effort="1" severity="optional">
```
Comment (Ladicek) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: I guess you might want to add "in RESTEasy 3" similarly to other rules.

File: rules-reviewed/eap7/eap6/resteasy.windup.xml [hunk 11]
```diff
@@ -0,0 +1,743 @@
...
+                    <tag>resteasy</tag>
+                </hint>
+            </perform>
+        </rule>
+        <rule id="resteasy-eap6-000004">
+            <when>
+                <javaclass references="org.jboss.resteasy.client.{type}" />
+            </when>
+            <perform>
+                <hint title="Deprecated Client API class org.jboss.resteasy.client.{type} in RESTEasy 3" effort="1" severity="optional">
```
Comment (Ladicek) [inline rules-reviewed/eap7/eap6/resteasy.windup.xml]: No other rule for Client API classes use "Deprecated Client API class", only "Deprecated class".

**Comment by Ladicek** [comment]
No other comments.

**Comment by mareknovotny** [comment]
:) new update

**Comment by Ladicek** [comment]
All my tests are passing. Good to merge from me.

**Comment by mareknovotny** [comment]
thanks @Ladicek!

I will let @jsight to review it and merge it.

**Pull request merged**